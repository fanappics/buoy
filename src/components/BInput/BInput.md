Buoy Input Component

```js
<b-input id="1" label="Example" />
```