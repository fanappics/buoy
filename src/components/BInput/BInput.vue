<template>
  <div class="flex">
    <label :for="id" :class="{ disabled: disabled }">
      {{ label }}<span v-if="required" aria-label="Required">*</span>
    </label>
    <div v-if="currency" :class="{ currency: true, focused: focused, disabled: disabled, invalid: invalid }" @click="$refs.input.focus()">
      <span style="font-weight: bold;">$</span>
      <input v-model="publicValue" v-bind="Object.assign(inputAttributes,validationAttributes)" v-validate.initial="validations" @focus="focused = true" @blur="focused = false" ref="input" @change="touched = true" />
    </div>
    <input v-else v-model="publicValue" v-bind="Object.assign(inputAttributes,validationAttributes)" v-validate.initial="validations" :class="{ invalid: invalid }" @change="touched = true" />
    <div v-if="invalid" :id="`error-${id}`" class="error">
      <span v-for="(error,index) in errors.all()" :key='index'>
        {{ error }}
      </span>
    </div>
  </div>
</template>

<script>
import validationMixIn from '../../mixins/validation'

export default {
  name: 'b-input',
  mixins: [validationMixIn({
    validateOn: 'blur|input'
  })],
  props: {
    //Required props
    id: {
      type: String,
      required: true
    },
    label: {
      type: String,
      required: true
    },
    //Optional props
    autofocus: Boolean,
    disabled: Boolean,
    max: Number,
    maxlength: Number,
    min: Number,
    minlength: Number,
    name: String,
    pattern: String,
    placeholder: String,
    readonly: Boolean,
    required: Boolean,
    type: {
      type: String,
      default: 'text',
      validator (value) {
        const validTypes = ['currency', 'date', 'datetime-local', 'email', 'month', 'number', 'password', 'search', 'tel', 'text', 'time', 'url', 'week']      
        if (validTypes.indexOf(value) < 0)
          throw new TypeError(`${value} is not a valid type for b-input`)
        return true
      }
    }
  },
  data () {
    return {
      focused: false,
    }
  },
  computed: {
    inputAttributes () {
      return {
        'aria-describedby': this.invalid ? `error-${this.id}` : '',
        'aria-label': this.label,
        'autofocus': this.autofocus,
        'disabled': this.disabled,
        'id': this.id,
        'name': this.name,
        'pattern': this.pattern,
        'placeholder': this.currency && !this.placeholder ? '0.00' : this.placeholder,
        'readonly': this.readonly,
        'required': this.required,
        'type': this.currency ? 'number' : this.type
      }
    },
    currency () {
      return this.type === 'currency'
    },
    //Creates an object that VeeValidate reads to apply certain rules
    validations () {
      let vals = {}
      if (this.max)
        vals.max_value = this.max
      if (this.maxlength)
        vals.max = this.maxlength
      if (this.min)
        vals.min_value = this.min
      if (this.minlength)
        vals.min = this.minlength
      if (this.pattern)
        vals.regex = this.pattern
      if (this.required)
        vals.required = true
      switch (this.type) {
        case 'currency':
          vals.decimal = 2
          break
        case 'email':
          vals.email = true
          break
        case 'tel':
          vals.digits = 10
          break
        case 'url':
          vals.url = true
      }
      return vals
    }
  }
}
</script>

<style scoped>
  input {
    border: solid 1px #dededf;
    border-radius: 4px;
    color: #333333;
    font-family: SFUIDisplay;
    font-size: 14px;
    padding: 12px;
  }
  input::-webkit-inner-spin-button {
    -webkit-appearance: none;
    -moz-appearance: none;
    margin: 0;
  }
  input::placeholder {
    color: #c1c1c1;
  }
  label {
    color: #333333;
    display: block;
    font-family: SFUIDisplay;
    font-size: 14px;
    font-weight: bold;
    padding-bottom: 6px;
  }
  .currency {
    align-items: baseline;
    border: solid 1px #dededf;
    border-radius: 4px;
    color: #333333;
    cursor: text;
    display: flex;
    font-family: SFUIDisplay;
    font-size: 14px;
    padding: 12px;
  }
  .currency input {
    background-color: transparent;
    border: 0;
    flex-grow: 1;
    padding: 0 0 0 12px;
  }
  .currency input:focus {
    outline: 0;
  }
  .disabled, input:disabled {
    opacity: 0.50;
  }
  .error {
    color: #d0021b;
    font-family: SFUIDisplay;
    font-size: 12px;
    padding-top: 5px;
  }
  .flex {
    display: flex;
    flex-direction: column;
  }
  .focused {
    outline: -webkit-focus-ring-color auto 5px;
    outline-offset: -2px;
  }
  .invalid {
    border: solid 1px #d0021b;
  }
</style>
