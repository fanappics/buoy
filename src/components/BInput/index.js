import BInput from './BInput.vue'

/* istanbul ignore next */
BInput.install = function install (Vue) {
  Vue.component(BInput.name, BInput)
}

export default BInput
