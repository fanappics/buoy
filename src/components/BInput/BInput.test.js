import { createLocalVue, shallow } from 'vue-test-utils'
import VeeValidate from 'vee-validate'
import BInput from './BInput'

const localVue = createLocalVue()
localVue.use(VeeValidate)

describe('Text input', () => {
  it('mounted html should match snapshot', () => {
    const wrapper = shallow(BInput, {
      propsData: {
        id: 'input-1',
        label: 'My Input',
        validationId: 'testValidationId'
      },
      localVue
    })
    expect(wrapper.html()).toMatchSnapshot()
  })
})

describe('Number input', () => {
  it('mounted html should match snapshot', () => {
    const wrapper = shallow(BInput, {
      propsData: {
        id: 'input-1',
        label: 'My Input',
        type: 'number',
        validationId: 'testValidationId'
      },
      localVue
    })
    expect(wrapper.html()).toMatchSnapshot()
  })
})

describe('Currency input', () => {
  it('mounted html should match snapshot', () => {
    const wrapper = shallow(BInput, {
      propsData: {
        id: 'input-1',
        label: 'My Input',
        type: 'currency',
        validationId: 'testValidationId'
      },
      localVue
    })
    expect(wrapper.html()).toMatchSnapshot()
  })
})

describe('Disabled input', () => {
  it('mounted html should match snapshot', () => {
    const wrapper = shallow(BInput, {
      propsData: {
        id: 'input-1',
        label: 'My Input',
        disabled: true,
        validationId: 'testValidationId'
      },
      localVue
    })
    expect(wrapper.html()).toMatchSnapshot()
  })
})

describe('Required input', () => {
  it('mounted html should match snapshot', () => {
    const wrapper = shallow(BInput, {
      propsData: {
        id: 'input-1',
        label: 'My Input',
        required: true,
        validationId: 'testValidationId'
      },
      localVue
    })
    expect(wrapper.html()).toMatchSnapshot()
  })
})

describe('Pattern input', () => {
  it('mounted html should match snapshot', () => {
    const wrapper = shallow(BInput, {
      propsData: {
        id: 'input-1',
        label: 'My Input',
        pattern: '^([0-9]+)$',
        validationId: 'testValidationId'
      },
      localVue
    })
    expect(wrapper.html()).toMatchSnapshot()
  })
})

describe('Readonly input', () => {
  it('mounted html should match snapshot', () => {
    const wrapper = shallow(BInput, {
      propsData: {
        id: 'input-1',
        label: 'My Input',
        readonly: true,
        validationId: 'testValidationId'
      },
      localVue
    })
    expect(wrapper.html()).toMatchSnapshot()
  })
})

describe('Named input', () => {
  it('mounted html should match snapshot', () => {
    const wrapper = shallow(BInput, {
      propsData: {
        id: 'input-1',
        label: 'My Input',
        name: 'Name of Input',
        validationId: 'testValidationId'
      },
      localVue
    })
    expect(wrapper.html()).toMatchSnapshot()
  })
})

describe('Autofocused input', () => {
  it('mounted html should match snapshot', () => {
    const wrapper = shallow(BInput, {
      propsData: {
        id: 'input-1',
        label: 'My Input',
        autofocus: true,
        validationId: 'testValidationId'
      },
      localVue
    })
    expect(wrapper.html()).toMatchSnapshot()
  })
})

describe('Required input failing validation', () => {
  it('mounted html should match snapshot', async () => {
    const wrapper = shallow(BInput, {
      propsData: {
        id: 'input-1',
        label: 'My Input',
        required: true,
        validationId: 'testValidationId'
      },
      localVue
    })

    const input = wrapper.find('input')
    input.trigger('focus')
    await wrapper.vm.$nextTick()
    input.element.value = ' '
    input.trigger('input')
    await wrapper.vm.$nextTick()

    expect(wrapper.html()).toMatchSnapshot()
  })
})
