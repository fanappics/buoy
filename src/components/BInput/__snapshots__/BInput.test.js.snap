// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Autofocused input mounted html should match snapshot 1`] = `
<div class="b-input b-flex-column">
    <label for="input-1" class="">
        My Input
        <!---->
    </label>
    <input aria-describedby="" aria-label="My Input" autofocus="autofocus" id="input-1" type="text" data-vv-as=" " data-vv-name="9" data-vv-scope="9" data-vv-validate-on="blur|input" class="" data-vv-id="9" aria-required="false" aria-invalid="false">
    <!---->
</div>
`;

exports[`Currency input mounted html should match snapshot 1`] = `
<div class="b-input b-flex-column">
    <label for="input-1" class="">
        My Input
        <!---->
    </label>
    <div class="b-input-el currency"><span style="font-weight: bold;">$</span>
        <input aria-describedby="" aria-label="My Input" id="input-1" placeholder="0.00" type="number" data-vv-as=" " data-vv-name="3" data-vv-scope="3" data-vv-validate-on="blur|input" data-vv-id="3" aria-required="false"
            aria-invalid="false">
    </div>
    <!---->
</div>
`;

exports[`Disabled input mounted html should match snapshot 1`] = `
<div class="b-input b-flex-column">
    <label for="input-1" class="b-disabled">
        My Input
        <!---->
    </label>
    <input aria-describedby="" aria-label="My Input" disabled="disabled" id="input-1" type="text" data-vv-as=" " data-vv-name="4" data-vv-scope="4" data-vv-validate-on="blur|input" class="" data-vv-id="4" aria-required="false" aria-invalid="false">
    <!---->
</div>
`;

exports[`Named input mounted html should match snapshot 1`] = `
<div class="b-input b-flex-column">
    <label for="input-1" class="">
        My Input
        <!---->
    </label>
    <input aria-describedby="" aria-label="My Input" id="input-1" name="Name of Input" type="text" data-vv-as=" " data-vv-name="8" data-vv-scope="8" data-vv-validate-on="blur|input" class="" data-vv-id="8" aria-required="false" aria-invalid="false">
    <!---->
</div>
`;

exports[`Number input mounted html should match snapshot 1`] = `
<div class="b-input b-flex-column">
    <label for="input-1" class="">
        My Input
        <!---->
    </label>
    <input aria-describedby="" aria-label="My Input" id="input-1" type="number" data-vv-as=" " data-vv-name="2" data-vv-scope="2" data-vv-validate-on="blur|input" class="" data-vv-id="2" aria-required="false" aria-invalid="false">
    <!---->
</div>
`;

exports[`Pattern input mounted html should match snapshot 1`] = `
<div class="b-input b-flex-column">
    <label for="input-1" class="">
        My Input
        <!---->
    </label>
    <input aria-describedby="" aria-label="My Input" id="input-1" pattern="^([0-9]+)$" type="text" data-vv-as=" " data-vv-name="6" data-vv-scope="6" data-vv-validate-on="blur|input" class="" data-vv-id="6" aria-required="false" aria-invalid="false">
    <!---->
</div>
`;

exports[`Readonly input mounted html should match snapshot 1`] = `
<div class="b-input b-flex-column">
    <label for="input-1" class="">
        My Input
        <!---->
    </label>
    <input aria-describedby="" aria-label="My Input" id="input-1" readonly="readonly" type="text" data-vv-as=" " data-vv-name="7" data-vv-scope="7" data-vv-validate-on="blur|input" class="" data-vv-id="7" aria-required="false" aria-invalid="false">
    <!---->
</div>
`;

exports[`Required input failing validation mounted html should match snapshot 1`] = `
<div class="b-input b-flex-column">
    <label for="input-1" class="">
        My Input<span aria-label="Required">*</span></label>
    <input aria-describedby="error-input-1" aria-label="My Input" id="input-1" required="required" type="text" data-vv-as=" " data-vv-name="10" data-vv-scope="10" data-vv-validate-on="blur|input"
        class="b-error-border" data-vv-id="10" aria-required="true" aria-invalid="true">
    <div id="error-input-1" class="b-error-text"><span>
      The   field is required.
    </span></div>
</div>
`;

exports[`Required input mounted html should match snapshot 1`] = `
<div class="b-input b-flex-column">
    <label for="input-1" class="">
        My Input<span aria-label="Required">*</span></label>
    <input aria-describedby="" aria-label="My Input" id="input-1" required="required" type="text" data-vv-as=" " data-vv-name="5" data-vv-scope="5" data-vv-validate-on="blur|input" class="" data-vv-id="5"
        aria-required="true" aria-invalid="false">
    <!---->
</div>
`;

exports[`Text input mounted html should match snapshot 1`] = `
<div class="b-input b-flex-column">
    <label for="input-1" class="">
        My Input
        <!---->
    </label>
    <input aria-describedby="" aria-label="My Input" id="input-1" type="text" data-vv-as=" " data-vv-name="1" data-vv-scope="1" data-vv-validate-on="blur|input" class="" data-vv-id="1" aria-required="false" aria-invalid="false">
    <!---->
</div>
`;
